### PR TITLE
Add new exit-code subcommand (AST-37832)

### PR DESCRIPTION
**Pull request by checkmarx-kobi-hagmi**
### Checklist

- [ ] I have added documentation for new/changed functionality in this PR (if applicable).
- [ ] I have updated the CLI help for new/changed functionality in this PR (if applicable).
- [ ] All active GitHub checks for tests, formatting, and security are passing
- [ ] The correct base branch is being used

### CODE REVIEW

File: internal/commands/result.go [hunk 1]
```diff
@@ -164,12 +164,32 @@ func NewResultsCommand(
...
 }
 
+func exitCodeSubCommand(scanWrapper wrappers.ScansWrapper) *cobra.Command {
+	exitCodeCmd := &cobra.Command{
+		Use:   "exit-code",
+		Short: "Get exit code and details of a scan",
+		Long:  "The exit-code command enables to get the exit code and failure details of a requested scan in Checkmarx One.",
+		Example: heredoc.Doc(
+			`
+			$ cx results exit-code --scan-id <scan Id> --scan-types <sast | iac-security>
```
Comment (checkmarx-kobi-hagmi) [inline internal/commands/result.go:181]: Check other types

File: internal/commands/result.go [hunk 2]
```diff
@@ -253,6 +273,86 @@ func resultBflSubCommand(bflWrapper wrappers.BflWrapper) *cobra.Command {
 	return resultBflCmd
 }
 
+func runGetExitCodeCommand(scanWrapper wrappers.ScansWrapper) func(cmd *cobra.Command, args []string) error {
+	return func(cmd *cobra.Command, args []string) error {
+		scanID, _ := cmd.Flags().GetString(commonParams.ScanIDFlag)
+		if scanID == "" {
+			return errors.New("scan ID is required")
```
Comment (checkmarx-kobi-hagmi) [inline internal/commands/result.go:280]: Extract to constant

File: internal/commands/result.go [hunk 3]
```diff
@@ -253,6 +273,86 @@ func resultBflSubCommand(bflWrapper wrappers.BflWrapper) *cobra.Command {
...
+			results = createAllFailedScannersResponse(scanResponseModel)
+			return printer.Print(cmd.OutOrStdout(), results, printer.FormatJSON)
+		}
+		scanTypes := sanitizeScannerNames(scanTypesFlagValue)
+		results = createRequestedScannersResponse(scanTypes, scanResponseModel)
+		return printer.Print(cmd.OutOrStdout(), results, printer.FormatJSON)
+	}
+}
+
+func createRequestedScannersResponse(scanTypes []string, scanResponseModel *wrappers.ScanResponseModel) []interface{} {
```
Comment (checkmarx-kobi-hagmi) [inline internal/commands/result.go:309]: check if possible to change to ScannerResponse

File: internal/commands/scan.go [hunk 4]
```diff
@@ -2000,13 +2000,49 @@ func isScanRunning(
...
+	}
+
+	return exitCodes.GeneralExitCode
+}
+
+const (
+	General = "general"
+	Sast    = "sast"
+	Sca     = "sca"
+	Iac     = "iac"
```
Comment (checkmarx-kobi-hagmi) [inline internal/commands/scan.go:2032]: check terminology - kics, iac, iac-security

File: internal/commands/result.go (hunk 1 above)
Comment (checkmarx-kobi-hagmi) [reply internal/commands/result.go:181]: discussed in product chatroom chat

File: internal/commands/result.go [hunk 5]
```diff
@@ -164,12 +164,32 @@ func NewResultsCommand(
...
-		showResultCmd, bflResultCmd, codeBashingCmd,
+		showResultCmd, bflResultCmd, codeBashingCmd, exitCodeSubcommand,
 	)
 	return resultCmd
 }
 
+func exitCodeSubCommand(scanWrapper wrappers.ScansWrapper) *cobra.Command {
+	exitCodeCmd := &cobra.Command{
+		Use:   "exit-code",
+		Short: "Get exit code and details of a scan",
```
[the remark this answers is not in the document]
Comment (checkmarx-kobi-hagmi) [reply internal/commands/result.go:178]: Long and short are confirmed by Eli.

File: internal/commands/result.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (checkmarx-kobi-hagmi) [reply internal/commands/result.go:182]: To be discussed in product channel. I followed the same pattern like we had in another command

File: internal/commands/result.go [hunk 6]
```diff
@@ -253,6 +273,86 @@ func resultBflSubCommand(bflWrapper wrappers.BflWrapper) *cobra.Command {
...
 
+func runGetExitCodeCommand(scanWrapper wrappers.ScansWrapper) func(cmd *cobra.Command, args []string) error {
+	return func(cmd *cobra.Command, args []string) error {
+		scanID, _ := cmd.Flags().GetString(commonParams.ScanIDFlag)
+		if scanID == "" {
+			return errors.New("scan ID is required")
+		}
+		scanResponseModel, errorModel, err := scanWrapper.GetByID(scanID)
+		if errorModel != nil || err != nil {
+			_ = printer.Print(cmd.OutOrStdout(), fmt.Sprintf("Scan %s not found", scanID), printer.FormatJSON)
```
[the remark this answers is not in the document]
Comment (checkmarx-kobi-hagmi) [reply internal/commands/result.go:285]: True.. Fixed!

File: internal/commands/result.go [hunk 7]
```diff
@@ -253,6 +273,86 @@ func resultBflSubCommand(bflWrapper wrappers.BflWrapper) *cobra.Command {
...
+				ScanID: scanResponseModel.ID,
+				Status: string(scanResponseModel.Status),
+			}
+			return printer.Print(cmd.OutOrStdout(), result, printer.FormatIndentedJSON)
+		}
+
+		var results []interface{}
+		scanTypesFlagValue, _ := cmd.Flags().GetString(commonParams.ScanTypes)
+		if scanTypesFlagValue == "" {
+			results = createAllFailedScannersResponse(scanResponseModel)
```
[the remark this answers is not in the document]
Comment (checkmarx-kobi-hagmi) [reply internal/commands/result.go:301]: According to the task if scan-types flag is not defined, it should return the error code and details for all the failed scanners. so if a scan is Completed this code will not be called and if a scan is Partial this function will be called and present all the failed ones.

File: internal/commands/result.go [hunk 8]
```diff
@@ -253,6 +273,86 @@ func resultBflSubCommand(bflWrapper wrappers.BflWrapper) *cobra.Command {
...
+			}
+		}
+	}
+	return results
+}
+
+func createAllFailedScannersResponse(scanResponseModel *wrappers.ScanResponseModel) []interface{} {
+	var results []interface{}
+	for i := range scanResponseModel.StatusDetails {
+		if scanResponseModel.StatusDetails[i].Status == wrappers.ScanFailed {
```
[the remark this answers is not in the document]
Comment (checkmarx-kobi-hagmi) [reply internal/commands/result.go:325]: Answered above.. let's discuss this.

File: internal/commands/result.go [hunk 9]
```diff
@@ -253,6 +273,86 @@ func resultBflSubCommand(bflWrapper wrappers.BflWrapper) *cobra.Command {
...
+		Details:   statusDetails.Details,
+		ErrorCode: stringifyErrorCode(statusDetails.ErrorCode),
+	}
+}
+
+func stringifyErrorCode(errorCode int) string {
+	if errorCode == 0 {
+		return ""
+	}
+	return strconv.Itoa(errorCode)
```
[the remark this answers is not in the document]
Comment (checkmarx-kobi-hagmi) [reply internal/commands/result.go:354]: If we fill 0 in the ErrorCode field in the json response we print to the console, it'll present this field... but I thought that if a scanner returns a success we don't want the error code field to be seen, right?

File: internal/commands/result_test.go [hunk 10]
```diff
@@ -35,6 +35,34 @@ func TestResultHelp(t *testing.T) {
...
 
+func TestResultsExitCode_OnFailedKicsScanner_PrintCorrectFailedScannerInfoToConsole(t *testing.T) {
+	execCmdNilAssertion(t, "results", "exit-code", "--scan-id", "fake-scan-id-kics-scanner-fail")
+}
+
+func TestResultsExitCode_OnFailedKicsAndScaScanners_PrintCorrectFailedScannersInfoToConsole(t *testing.T) {
+	execCmdNilAssertion(t, "results", "exit-code", "--scan-id", "fake-scan-id-multiple-scanner-fails")
+}
+
+func TestResultsExitCode_OnFailedKicsAndScaScannersAndRequestedScannerIsSca_PrintCorrectFailedScannersInfoToConsole(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (checkmarx-kobi-hagmi) [reply internal/commands/result_test.go:46]: There's no trivial way to read content from a console. I can try returning the json, reading it into a buffer somehow, but I'll have to dig into this

[the remark this answers is not in the document]
Comment (checkmarx-kobi-hagmi) [reply internal/commands/result_test.go:46]: Try to write shorter test method names

File: internal/commands/result.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (OrShamirCM) [reply internal/commands/result.go]: As discussed, please add all engines that we support

[the remark this answers is not in the document]
Comment (checkmarx-kobi-hagmi) [reply internal/commands/result.go]: it's updated already:
cx results exit-code --scan-id <scan Id> --scan-types <sast | sca | iac-security | apisec>

File: internal/commands/result.go [hunk 11]
```diff
@@ -164,12 +165,32 @@ func NewResultsCommand(
...
 }
 
+func exitCodeSubCommand(scanWrapper wrappers.ScansWrapper) *cobra.Command {
+	exitCodeCmd := &cobra.Command{
+		Use:   "exit-code",
+		Short: "Get exit code and details of a scan",
+		Long:  "The exit-code command enables you to get the exit code and failure details of a requested scan in Checkmarx One.",
+		Example: heredoc.Doc(
+			`
+			$ cx results exit-code --scan-id <scan Id> --scan-types <sast | sca | iac-security | apisec>
```
Comment (OrShamirCM) [inline internal/commands/result.go:182]: now I see, we are not supporting containers?

Comment (checkmarx-kobi-hagmi) [reply internal/commands/result.go]: Added now

File: internal/commands/result.go [hunk 12]
```diff
@@ -164,12 +165,32 @@ func NewResultsCommand(
...
 }
 
+func exitCodeSubCommand(scanWrapper wrappers.ScansWrapper) *cobra.Command {
+	exitCodeCmd := &cobra.Command{
+		Use:   "exit-code",
+		Short: "Get exit code and details of a scan",
+		Long:  "The exit-code command enables you to get the exit code and failure details of a requested scan in Checkmarx One.",
+		Example: heredoc.Doc(
+			`
+			$ cx results exit-code --scan-id <scan Id> --scan-types <sast | sca | iac-security | apisec | containers>
```
Comment (OrShamirCM) [inline internal/commands/result.go:182]: we are not supporting containers

Comment (checkmarx-kobi-hagmi) [reply internal/commands/result.go]: revmoved

File: internal/commands/result.go [hunk 13]
```diff
@@ -253,6 +274,104 @@ func resultBflSubCommand(bflWrapper wrappers.BflWrapper) *cobra.Command {
...
+		scanTypes := sanitizeScannerNames(scanTypesFlagValue)
+		results = createRequestedScannersResponse(scanTypes, scanResponseModel)
+	}
+
+	return results
+}
+
+func createRequestedScannersResponse(scanTypes map[string]string, scanResponseModel *wrappers.ScanResponseModel) []ScannerResponse {
+	var results []ScannerResponse
+	for i, StatusDetail := range scanResponseModel.StatusDetails {
```
Comment (OrShamirCM) [inline internal/commands/result.go:330]: use either  or status detail

Comment (checkmarx-kobi-hagmi) [reply internal/commands/result.go]: fixed

File: test/integration/result_test.go [hunk 14]
```diff
@@ -21,7 +22,30 @@ const (
...
+		"results", "exit-code",
+		flag(params.ScanIDFlag), scanID,
+		flag(params.ScanTypes), "sast,sca",
+	}
+
+	err, _ := executeCommand(t, args...)
+	assert.NilError(t, err)
+}
+
+func TestResultsExitCode_NoScanIdSent_FailCommandWithError(t *testing.T) {
```
Comment (OrShamirCM) [inline test/integration/result_test.go:38]: have integration test that assert the repsonse

File: internal/commands/result.go [hunk 15]
```diff
@@ -343,7 +343,7 @@ func createRequestedScannersResponse(scanTypes map[string]string, scanResponseMo
 func createAllFailedScannersResponse(scanResponseModel *wrappers.ScanResponseModel) []ScannerResponse {
 	var results []ScannerResponse
 	for i := range scanResponseModel.StatusDetails {
-		if scanResponseModel.StatusDetails[i].Status == wrappers.ScanFailed && scanResponseModel.StatusDetails[i].Name != General {
+		if scanResponseModel.StatusDetails[i].Status == wrappers.ScanFailed {
```
Comment (OrShamirCM) [inline internal/commands/result.go:346]: so we will always add the general and it's approved?

Comment (OrShamirCM) [reply internal/commands/result.go:346]: can you have a test for this too?

File: internal/commands/scan.go [hunk 16]
```diff
@@ -2068,13 +2069,49 @@ func isScanRunning(
...
+	return exitCodes.MultipleEnginesFailedExitCode
+}
+
+const (
+	General     = "general"
+	Sast        = "sast"
+	Sca         = "sca"
+	IacSecurity = "iac-security" // We get 'kics' from AST. Added for forward compatibility
+	Kics        = "kics"
+	APISec      = "apisec"
```
Comment (OrShamirCM) [inline internal/commands/scan.go:2103]: this is the correct name? not api-security?

**Review by OrShamirCM — APPROVED** [review]

File: internal/commands/scan.go (hunk 16 above)
Comment (checkmarx-kobi-hagmi) [reply internal/commands/scan.go:2103]: image

File: internal/commands/result.go (hunk 15 above)
Comment (checkmarx-kobi-hagmi) [reply internal/commands/result.go:346]: Yes, because we need to display issues that are not engine-specific.
tested were added in unit tests.

**Pull request merged**